docs/01-tooling/03-fast-and-isolated-JS-unit-tests.md

```diff
@@ -2,10 +2,159 @@
+coverage. We added another script -[`coverage`][coverage]- which just run the
+test command, but with the code coverage tool [`istanbul`][istanbul] first, and
+just like that we got back coverage reports for the frontend code.
+
+We recommend this approach for others wanting to improve how they test, and
+would be happy to help you figure out if this approach would work for you. For
+example, you can use this CLI runner, even if your JS sources just use globals
+instead of common.js or ES modules.
+
 ### Problems
+
+Overall, the move has gone great and we don't many issues to report.
+
+When migrating existing tests, it is sometimes a bit tricky to figure out how to
+move them to the isolated environment, since most of the [MediaWiki JS
+library][mw] is not available as an npm package, so in some occasions we had to
+restructure the code a bit to not implicitly assume as much of MediaWiki being
+available, and other times we had to set up some stubs for the tests to run
+well. This had the added benefit that the dependency on MediaWiki core libraries
+is explicit in the tests, so we should notice when adding new dependencies or
+changing them because of the failing tests, keeping the behavior and
+dependencies explicit.
+
+Another extra thing that we have been doing has been maintaining
+[`mw-node-qunit`][mw-node-qunit], which has taken a bit of additional time.
+Making sure our wrapper works well with qunitjs, and updating the dependency
+versions to not fall behind and leverage improvements on the libraries.
+
+We also expect some more work on the wrapper as [QUnit Plugins (or "Setup Step")
+in CLI \#1221][1221] gets worked on, or when [CLI: Implement --require option
+\#1271][1271] gets released, to simplify our implementation and be able to
+leverage the superior QUnit.js tap output and the rest of the CLI options like
+filters. The good thing is that the CLI wrapper should be useful for any
+Wikimedia projects as is.
+
+We will also be looking into moving the repository to the Wikimedia organization
+in github if other teams or projects adopt this testing strategy.
+
+[adr-7]: https://github.com/wikimedia/mediawiki-extensions-Popups/blob/2ddf8a96d8df27d6b5e8b4dd8ef33581951db9fe/doc/adr/0007-prefer-running-qunit-tests-in-node-js.md
+[qunit]: https://www.npmjs.com/package/qunitjs
+[jquery]: https://www.npmjs.com/package/jquery
+[sinon]: https://www.npmjs.com/package/sinon
+[jsdom]: https://www.npmjs.com/package/jsdom
+[mw-node-qunit]: https://github.com/joakin/mw-node-qunit/
+[stubs]: https://github.com/wikimedia/mediawiki-extensions-Popups/blob/2ddf8a96d8df27d6b5e8b4dd8ef33581951db9fe/tests/node-qunit/stubs.js
+[coverage]: https://github.com/wikimedia/mediawiki-extensions-Popups/blob/2ddf8a96d8df27d6b5e8b4dd8ef33581951db9fe/package.json#L12
+[istanbul]: https://istanbul.js.org/
+[mw]: https://doc.wikimedia.org/mediawiki-core/master/js/
+[1221]: https://github.com/qunitjs/qunit/issues/1221
+[1271]: https://github.com/qunitjs/qunit/pull/1271
```

Comment:
This article as a whole touches on many truths of testing. I would really love to see it boiled down in the end into a summary with a bulleted list of guidelines with possible references to some of the talks and learnings materials you're always sharing. Here's my incomplete summary:

- Untested code is unmaintained.
- Tests that run slow are never run.
- Monolithic integration tests are slow to write, read, modify, debug, and execute; isolated unit tests are the opposite.
- Code that is difficult to test in isolation may be indicative of functional concerns.
- Efficient tests = efficient development.

Reply:
That is a great list. I've added it to the requirements in the form of "Given that {your list}, we need {requirements}"

In the form of summary there is a conclusion section now that is more positive and uplifting